src/modules/ekf2/EKF/gps_control.cpp

```diff
@@ -52,22 +52,42 @@ void Ekf::controlGpsFusion()
 		const gpsSample &gps_sample{_gps_sample_delayed};
 
 		updateGpsYaw(gps_sample);
-		updateGpsVel(gps_sample);
-		updateGpsPos(gps_sample);
 
 		const bool gps_checks_passing = isTimedOut(_last_gps_fail_us, (uint64_t)5e6);
 		const bool gps_checks_failing = isTimedOut(_last_gps_pass_us, (uint64_t)5e6);
 
 		controlGpsYawFusion(gps_sample, gps_checks_passing, gps_checks_failing);
 
+		// GNSS velocity
+		const Vector3f velocity{gps_sample.vel};
+		const float vel_var = sq(gps_sample.sacc);
+		const Vector3f vel_obs_var{vel_var, vel_var, vel_var * sq(1.5f)};
+		updateVelocityAidSrcStatus(gps_sample.time_us, velocity, vel_obs_var, fmaxf(_params.gps_vel_innov_gate, 1.f), _aid_src_gnss_vel);
+		_aid_src_gnss_vel.fusion_enabled = (_params.gnss_ctrl & GnssCtrl::VEL);
+
+		// GNSS position
+		const float pos_var_lower_limit = fmaxf(_params.gps_pos_noise, 0.01f);
+		float pos_var = sq(fmaxf(gps_sample.hacc, pos_var_lower_limit));
+
+		if (!isOtherSourceOfHorizontalAidingThan(_control_status.flags.gps)) {
+			// if we are not using another source of aiding, then we are reliant on the GPS
+			// observations to constrain attitude errors and must limit the observation noise value.
+			float upper_limit = fmaxf(_params.pos_noaid_noise, pos_var_lower_limit);
+			pos_var = fminf(pos_var, upper_limit);
+		}
+
+		updateHorizontalPositionAidSrcStatus(gps_sample.time_us, gps_sample.pos, Vector2f(pos_var, pos_var), fmaxf(_params.gps_pos_innov_gate, 1.f), _aid_src_gnss_pos);
+		_aid_src_gnss_pos.fusion_enabled = (_params.gnss_ctrl & GnssCtrl::HPOS);
+
 		// update GSF yaw estimator velocity (basic sanity check on GNSS velocity data)
 		if (gps_checks_passing && !gps_checks_failing) {
-			_yawEstimator.setVelocity(_gps_sample_delayed.vel.xy(), _gps_sample_delayed.sacc);
+			_yawEstimator.setVelocity(velocity.xy(), gps_sample.sacc);
 		}
 
 		// Determine if we should use GPS aiding for velocity and horizontal position
 		// To start using GPS we need angular alignment completed, the local NED origin set and GPS data that has not failed checks recently
-		const bool mandatory_conditions_passing = _control_status.flags.tilt_align
+		const bool mandatory_conditions_passing = ((_params.gnss_ctrl & GnssCtrl::HPOS) || (_params.gnss_ctrl & GnssCtrl::VEL))
```

Comment:
We'll go through the early stop and return at the top of the function if none of those gnss ctrl flags are set; do you plan to remove it?

Reply:
I had a bigger follow up change where I started removing the checks at the top, but maybe we should talk about overall desired behaviour first.